content/posts/2018-10-30_Docker-dev-and-prod-ready-setup-for-SPA-app.md

```diff
@@ -0,0 +1,435 @@
+
+Be more light and flexible, especially on the development environment, when you need this a lot with continuous changes you're making.
+
+### <a name="multi-compose-files"></a>Multiple compose files
+
+Firstly, what is `multiple compose files`?
+By default, our docker-compose will read base `docker-compose.yml` file wherein the convention, it should be our base configuration, which **will be shared over all environments**.
+After that, compose will append the final docker-compose file with override file like `docker-compose.override.yml`.
+When some services are defined in both, they will be merged together.
+This feature is really cool and will help us to keep our setup clear and easy to maintain.
+
+So, we'll create `docker-compose.yml` which will our base for all environments, also production etc. 
+
+```yml
+# docker-compose.yml
+version: '3.6'
+
+services:
+  client:
+    env_file:
+      - .env
+```
+
+So, what happened here? Probably you know `docker-compose` schema and properties, but if not, I will explain a little:
+
+* `version: 3.6` - Shows supported compose file version to `docker-compose`, so we'll know what syntax and features we could use.
+* `services:` - main compose purpose, registering services to run them all neither using a command for one by one like `docker run ...`
+* `client:` - registering our first and main service - `client` which will be our React app,
+* `env_file:` - pointing the `.env` file which should be used to get global environments values for our setup. This file helps us to storage variables for our app like exposed port, application environment, compose project name and much more, that we could easily change due to fact that they are in one place.
+
+Okay, now our **Docker** setup contains only base `docker-compose.yml`, so we need to add first override compose file for development environment: `docker-compose.dev.yml`
+
+```yml
+# docker-compose.dev.yml
+version: '3.6'
+
+services:
+  client:
+    image: node:carbon
+    volumes:
+      - ..:/opt/app:cached
+    ports:
+      - "${NODE_PORT}:${NODE_PORT}"
+    working_dir: /opt/app
+    command: "/bin/sh -c 'yarn install --frozen-lockfile --production=false; yarn start'"
+```
+
+Describe a bit more this little magic:
+
+* `image:` - here we're specifying that we want to use an image with predefined `node.js` in `carbon` (`8.x`) version. Based on this image, Docker will build a container for our application (`client`),
+* `volumes:` - this allows us to mount some data volume (we needn't copy data, because we mount data from our host to container) as `HOST:CONTAINER` path. So we're mounting `..` from the host (project root directory) to `/opt/app` in our container. We also add `:cached` to improving Mac's performance. More details about it [here](https://docs.docker.com/docker-for-mac/osxfs-caching/#cached).
```

Comment:
```suggestion
* `volumes:` - this allows us to mount some data volume (we don't need to copy data, because we mount data from our host to container) as `HOST:CONTAINER` path. So we're mounting `..` from the host (project root directory) to `/opt/app` in our container. We also add `:cached` to improving Mac's performance. More details about it [here](https://docs.docker.com/docker-for-mac/osxfs-caching/#cached).
```